source/connection_monitor.c

```diff
@@ -0,0 +1,222 @@
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/http/private/connection_monitor.h>
+
+#include <aws/http/connection.h>
+#include <aws/http/statistics.h>
+#include <aws/io/channel.h>
+#include <aws/io/logging.h>
+#include <aws/io/statistics.h>
+
+#include <aws/common/clock.h>
+
+#include <inttypes.h>
+
+static void s_process_statistics(
+    struct aws_crt_statistics_handler *handler,
+    struct aws_crt_statistics_sample_interval *interval,
+    struct aws_array_list *stats_list,
+    void *context) {
+
+    (void)interval;
+
+    struct aws_statistics_handler_http_connection_monitor_impl *impl = handler->impl;
+    if (!aws_http_connection_monitoring_options_is_valid(&impl->options)) {
+        return;
+    }
+
+    uint64_t pending_read_interval_ms = 0;
+    uint64_t pending_write_interval_ms = 0;
+    uint64_t bytes_read = 0;
+    uint64_t bytes_written = 0;
+    uint32_t current_outgoing_stream_id = 0;
+    uint32_t current_incoming_stream_id = 0;
+
+    /*
+     * Pull out the data needed to perform the throughput calculation
+     */
+    size_t stats_count = aws_array_list_length(stats_list);
+    for (size_t i = 0; i < stats_count; ++i) {
+        struct aws_crt_statistics_base *stats_base = NULL;
+        if (aws_array_list_get_at(stats_list, &stats_base, i)) {
+            continue;
+        }
+
+        switch (stats_base->category) {
+            case AWSCRT_STAT_CAT_SOCKET: {
+                struct aws_crt_statistics_socket *socket_stats = (struct aws_crt_statistics_socket *)stats_base;
+                bytes_read = socket_stats->bytes_read;
+                bytes_written = socket_stats->bytes_written;
+                break;
+            }
+
+            case AWSCRT_STAT_CAT_HTTP1_CHANNEL: {
+                struct aws_crt_statistics_http1_channel *http1_stats =
+                    (struct aws_crt_statistics_http1_channel *)stats_base;
+                pending_read_interval_ms = http1_stats->pending_incoming_stream_ms;
+                pending_write_interval_ms = http1_stats->pending_outgoing_stream_ms;
+                current_outgoing_stream_id = http1_stats->current_outgoing_stream_id;
+                current_incoming_stream_id = http1_stats->current_incoming_stream_id;
+
+                break;
+            }
+
+            default:
+                break;
+        }
+    }
+
+    struct aws_channel *channel = context;
+
+    uint64_t bytes_per_second = 0;
+    uint64_t max_pending_io_interval_ms = 0;
+
+    if (pending_write_interval_ms > 0) {
+        double fractional_bytes_written_per_second =
+            (double)bytes_written * (double)AWS_TIMESTAMP_MILLIS / (double)pending_write_interval_ms;
```

Comment:
I think you should replace `AWS_TIMESTAMP_MILLIS` here with MY_REPORT_INTERVAL_MS?

`AWS_TIMESTAMP_MILLIS` happens to be 1000, and `s_get_report_interval_ms()` has `return 1000;`
but using the same constant/define in both places would be preferable.

Assuming my understanding of this math is correct

Or should we measure the ACTUAL time elapsed since the last sample and use that here, instead of a hardcoded number?

Reply:
We're using the actual time we were "trying" to write (with the caveat that, as you pointed out elsewhere, it's not completely accurate to the socket level).  The stats reporting interval is completely independent and not a factor in this calculation.